compiler/rustc_metadata/src/native_libs.rs

```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
             }
         }
     }
+
+    fn i686_arg_list_size(&self, item: &hir::ForeignItemRef<'_>) -> usize {
+        let argument_types: &List<Ty<'_>> = self.tcx.erase_late_bound_regions(
+            self.tcx
+                .type_of(item.id.def_id)
+                .fn_sig(self.tcx)
+                .inputs()
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        // stdcall on i686 never allocates less than 32 bits for an argument.
```

Comment:
Is this true for fastcall as well?

(Also I suspect that we already have the `i686_arg_list_size` logic somewhere in rustc, but can't check right now.)

Reply:
It appears to be true: running run-make/raw-dylib-stdcall (replacing `stdcall` with `fastcall` throughout) does work.  As per comment below, I'll add tests for this case, and I'll update the comment.

And I wouldn't be at all surprised to learn that `i686_arg_list_size` is duplicate logic, but after digging around myself and consulting with various people, I was unable to find existing code that does the same thing.  If we can find the existing function, I'll be more than happy to use that instead.

Reply:
The closest logic that I could find is `compiler\rustc_target\src\abi\call\x86.rs`, but it looks like the mangling logic doesn't care what is passed in registers or somewhere else, it just sums up the sizes.

Reply:
You are correct that the mangling logic doesn't differentiate between arguments passed on the stack and arguments passed in registers.  I haven't seen that documented anywhere, but the fact that this PR's run-make test passes provides pretty solid evidence for it.  😃 

I have a vague recollection that the stdcall convention requires the caller to allocate but not initialize stack space for arguments passed in registers, in order to give the called function a place to spill those registers if necessary, but I could be dreaming that up.  Either way, I think we've addressed this point -- if we find a better pre-existing function, then we can start another conversation and I'll address it then.

Reply:
Ugh, I looked through some code in LLVM and found at least two instances of this logic:
- `addByteCountSuffix` in `llvm-project\llvm\lib\IR\Mangler.cpp`  - aligns to 32 bit (or pointer size in general) instead of using it as a minimum like this PR, comment (incorrectly?) says that the result is a number of words rather than the number of bytes.
- `MangleContext::mangleName` in `llvm-project\clang\lib\AST\Mangle.cpp` - seems to do the same thing, but with some messy divisions and multiplications to get the number of bytes (again) instead of the number of words.

Looks like the alignment vs minimum mismatch is a thing to fix.

Reply:
Yeah, I saw the comment in `addByteCountSuffix`, or one just like it.  I can only assume, based on experience, that the use of "words" there is incorrect -- using size in words instead of bytes in the mangled names results in an import library that doesn't link correctly against the DLL produced by MSVC.  The details of stdcall that I've been able to find at docs.microsoft.com are suspiciously silent on this point, so I kinda have to take the empirical results as definitive.  I'd personally prefer to have better clarity in the spec on this point, but 🤷 .

Reply:
> Looks like the alignment vs minimum mismatch is a thing to fix.

Well, after further experimentation, I agree that `max(4, layout.size.bytes_usize())` is not correct, and I've updated this, although I haven't pushed yet.

Based on what I've observed, it seems that stdcall expects that the space for a given argument is the smallest multiple of 4 bytes into which that argument can fit; this is true even for types like `f64` that might under other circumstances be aligned to the nearest 8-byte boundary.

The alignment information that I get from the `Layout` struct doesn't really help here.  As one example, the `Layout` value for `u8` indicates an alignment of 1 byte (both abi and preferred), which doesn't match stdcall: `fn f(x: u8, y: u8)` has an argument size of 8 bytes, not 2.

Unfortunately, messy code is probably unavoidable -- best I've got so far is `(layout.size.bytes_usize() + 3) & !3`, with a comment explaining what's going on.

Beyond that, is there anything else you want to see here?

Reply:
>Beyond that, is there anything else you want to see here?

No, the `min` was the only thing that looked incorrect here.